### ISSUE TITLE
Latency issues on Linux

### ISSUE THREAD

**Issue by collin-kemper** [frewsxcv/rust-dark-light#17]
There are significant latency issues when calling `detect` on Arch Linux using i3. D-bus times out waiting for freedesktop to respond, which results in a 25 second latency.

**Comment by Be-ing** [comment]
Do you have an XDG Desktop Portal implementation installed?

**Comment by collin-kemper** [comment]
Yes, I have [XDG Desktop Portal](https://archlinux.org/packages/extra/x86_64/xdg-desktop-portal/) installed.

**Comment by Be-ing** [comment]
xdg-desktop-portal doesn't do anything by itself; you need one of the backend implementations (xdg-desktop-portal-gtk or xdg-desktop-portal-kde).

**Comment by collin-kemper** [comment]
I also have xdg-desktop-portal-gtk installed. I also installed xdg-desktop-portal-kde and this did not change the latency.

**Comment by Be-ing** [comment]
> I start i3 using xinit.

That's likely the problem. Normally when logging in with a display manager, systemd starts the xdg-desktop-portal service.

**Comment by collin-kemper** [comment]
Is it possible to support graceful failure in the event of xdg-desktop-portal not being started? I'm not familiar enough with d-bus to know whether this is possible.

**Comment by Be-ing** [comment]
zbus doesn't have a convenient API to set a maximum timeout, but it [looks like](https://docs.rs/zbus/latest/zbus/blocking/struct.Connection.html#method.monitor_activity) that is possible.

**Comment by Be-ing** [comment]
There is already a [zbus issue](https://gitlab.freedesktop.org/dbus/zbus/-/issues/103) for that.

**Comment by Be-ing** [comment]
This could be alleviated by switching to an async API: https://github.com/frewsxcv/rust-dark-light/pull/10#pullrequestreview-854857594=

**Comment by zeenix** [comment]
zbus author here. :wave: 

> This could be alleviated by switching to an async API: [#10 (review)](https://github.com/frewsxcv/rust-dark-light/pull/10#pullrequestreview-854857594)

Correct. Runtimes provide [timeout API](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) already and since the async API is the recommended one, the priority for adding this feature got lowered with 2.0 when we switched to async being the primary API in 2.0.

**Comment by collin-kemper** [comment]
My vote would be for adding an async API. I briefly looked over the PR - is there a way I can contribute to it?

**Comment by edfloreshz** [comment]
We considered splitting the API and supporting both, but there's simply no way to fetch the mode in macOS and Windows asynchronously, the only platform that supports that is FreeDesktop via `ashpd`.

We opted instead for keeping the library sync and using `futures_lite::future::block_on()`, we could add a timeout using `async-std` to avoid latency issues.

The other option would be to switch the function signature to `async fn` but every platform besides FreeDesktop wouldn't be doing any async work.

**Comment by edfloreshz** [comment]
That sounds good to me.

**Comment by edfloreshz** [comment]
Yeah, I was thinking 20-30 would be enough.

**Comment by castarco** [comment]
2026 here. Latency is still a problem in some Linux systems. Having 25ms hardcoded in there is far from ideal in this situation.

I'm using `mundy` instead, where I can set my own timeout, and 75ms works much better.
